app/src/main/java/com/poupa/vinylmusicplayer/provider/MusicPlaybackQueueStore.java

```diff
@@ -208,18 +210,45 @@ private synchronized void saveQueue(final String tableName, @NonNull final Array
 
 
     @NonNull
-    private ArrayList<IndexedSong> getSongPosition(@Nullable Cursor cursor, @NonNull final ArrayList<Song> songs) {
+    private ArrayList<IndexedSong> getSongPosition(@Nullable Cursor cursor, @NonNull final ArrayList<Song> songs, @NonNull final ArrayList<Long> removedSongIds) {
         ArrayList<IndexedSong> queue = new ArrayList<>();
+        ArrayList<Integer> removedIndexes = new ArrayList<>();
 
         if (cursor != null && cursor.moveToFirst()) {
             int i = 0;
-            int idColumns = cursor.getColumnIndex(MusicPlaybackColumns.INDEX_IN_QUEUE);
+            int idColumn = cursor.getColumnIndex(BaseColumns._ID);
+            int indexColumn = cursor.getColumnIndex(MusicPlaybackColumns.INDEX_IN_QUEUE);
+
             do {
-                queue.add(new IndexedSong(songs.get(i), cursor.getInt(idColumns), IndexedSong.INVALID_INDEX));
+                long songId = cursor.getLong(idColumn);
+                int songIndex = cursor.getInt(indexColumn);
+
+                if (removedSongIds.contains(songId)) {
+                    // Note that the queue store will still contain the orphan songs
+                    // It will only gets cleaned up when a new queue is created
+                    removedIndexes.add(songIndex);
+                    continue;
+                }
+
+                queue.add(new IndexedSong(songs.get(i), songIndex, IndexedSong.INVALID_INDEX));
                 i++;
             } while (cursor.moveToNext());
         }
 
+        // Adjust the index value, since there are removed songs
+        if (!removedIndexes.isEmpty())
+        {
+            Collections.sort(removedIndexes);
+            Collections.reverse(removedIndexes);
```

Comment:
Why do you do a Collections.reverse and not just a reversed for loop ?
for (int i = removedIndexes.size() - 1; i >= 0; i--) {
   ...
   removedIndexes.get(i)
   ...
}

Reply:
except for this, the rest of the proposed changed seems logical and seems to worked after some small test

Reply:
I find it easier to read the syntax `for ( iter : collection)`. I've changed however to merge the sort and resever operation in one line